### PR TITLE
build(client): add vue-composition api, update vue-apollo, update vuelidate

### LINKED ISSUE #402: Add composable for checking abilities and roles of the current user
### Problem/Why
Checking the abilities or roles of a user is a common template requirement for showing and hiding UI elements. 

### Goal/What
A globally available a composable API that standardizes the ability of a template developer to check abilities and/or roles.

### Solution/How
Implement a currentUser composable API. The api should fetch the currentuser query from apolloclient and create computed functions for can(ability) and hasRole(role).  These functions should return:
- can('edit:User') .. with abilities,
  - `['*']: true`
  - `['edit:User', ...]: true`
  - `['edit:Foo', ...]: false`
  - `[]: false`
- hasRole('admin') .. with roles
  - `[]: false`
  - `['admin', ...]: true`
- hasRole() .. with roles (equates to has ANY role)
  - `[]: false`
  - `['editor', ...]: true`
  - `['admin', ...]: true`

### To Do / Acceptance Criteria

- [x]  Implement composable as described.
- [x]  Add jest tests to confirm functionality

### Review Checklist

- [ ] Tests are passing and complete.

### Accessibility Checklist

No accessibility concerns

### LINKED ISSUE #412: Update vue-apollo to version 4
### Problem/Why
The current version of vue-apollo does not provide the composition API. :-( The composition API will help make portions of our code more reuseable.

### Goal/What
Provide developer access to the composition API. 

### Solution/How

Remove quasar-app-apollo and integrate vue-apollo@4 directly.  vue-apollo@4 is compatible with Vue v2.  quasar-app-apollo@next provides vue-apollo@4, but requires quasar@v2 which we aren't ready to upgrade to.

### To Do / Acceptance Criteria

- [x] Remove `@quasar/apollo` extension using the quasar-cli
- [x] Install `graphql graphql-tag and @apollo/client`
- [x] Follow [options API installation docs](https://v4.apollo.vuejs.org/guide-option/setup.html#_1-install-vue-apollo-option) to ensure compatibility with out current code.
- [x] Follow [composable-api installation docs](https://v4.apollo.vuejs.org/guide-composable/setup.html#_1-install-vue-apollo-composable)
- [x] Ensure existing tests pass.
- [x] Update `page/admin/UsersIndex.vue` to utilize the composition API
- [x] Update tests as needed to pass

### Review Checklist

- [ ] All tests pass
- [ ] Code review

### Accessibility Checklist

No accessibility issues.

### LINKED ISSUE #413: Install vue/composition-api
### Problem/Why
There are challenges and drawbacks with re-using code with the traditional approaches available in Vue 2 (i.e. mixins, mixin factories, scoped slots).

### Goal/What
Implement the vue2 composition api to improve the reuse of code in an organized, maintainable way.  

### Solution/How
The vue2 composition api is compatible with our current quasar.  Install and configure the composition api.

### To Do / Acceptance Criteria

- [x] Install the @vue/composition-api: https://github.com/vuejs/composition-api
- [x] When instructed to add Vue.use(VueCompositionApi) read [QuasarCLI/boot files](https://v1.quasar.dev/quasar-cli/boot-files#introduction)
- [x] Ensure that all tests remain passing.
- [x] Ensure existing tests pass.
### Review Checklist

- [ ] Existing tests pass

### Accessibility Checklist

No accessibility concerns.

### LINKED ISSUE #414: Update vuelidate to version 2
### Problem/Why
Vuelidate v0 does not make use of the composition API and it is therefore more cumbersome to reuse validators/validation logic.

### Goal/What
Update vuelidate to the v2 beta

### To Do / Acceptance Criteria

- [x] Remove the existing vuelidate plugin.
- [x] Install per the (documentation)[https://vuelidate-next.netlify.app/#installation].  Note that boot files should not be needed as vuelidate v2 does not use a plugin.
- [x] Update existing uses of validationMixin following the [migration guide](https://vuelidate-next.netlify.app/migration_guide.html)
- [x] Ensure existing tests pass.

### Review Checklist

- [ ] Tests pass
- [ ] Code review

### Accessibility Checklist

No accessibility issues.

### PR DESCRIPTION
**Pull request by wreality**
closes #412, closes #413, closes #414, closes #402

### CODE REVIEW

**Comment by wreality** [comment]
@gmeben Ignore the review request.  This isnt quite ready yet.

**Comment by wreality** [comment]
@gmeben or @Sruthin86 this should be ready for review now.

**Comment by gmeben** [comment]
I will do my best to get this reviewed, but I may struggle to submit a final review today due to the size of this PR

**Comment by wreality** [comment]
No problem. It's a big one fo sho.

**Comment by wreality** [comment]
@gmeben Also if you want to talk through any of it, I'm happy to jump on a call.

File: client/src/pages/Admin/Publications.vue [hunk 1]
```diff
@@ -14,18 +14,18 @@
...
-            class="form-error text-white bg-negative text-center q-mt-xs"
-            data-cy="banner_form_error"
-            v-text="$t(`publications.create.failure`)"
-          />
+          >
+            <error-field-renderer
+              :errors="$v.new_publication.name.$errors"
+              prefix="publications.create.name"
+              data-cy="name_field_error"
+            />
```
Comment (gmeben) [inline client/src/pages/Admin/Publications.vue:26]: It seems the convention in Quasar by default is to display one error message at a time. I'm curious what the reasoning is to deviate from that and show multiple error messages at once. I can imagine there's advantages to both approaches but I don't think I fully understand all of the reasoning behind this choice.

Also, all of the [examples of QInputs with nested content in Quasar's documentation](https://quasar.dev/vue-components/input) shows them using a `<template>` tag with a `slot` attribute  to dictate the position of the nested content relative to the input.

```suggestion
            <template #append>
              <error-field-renderer
                :errors="$v.new_publication.name.$errors"
                prefix="publications.create.name"
                data-cy="name_field_error"
              />
            </template>
```

File: client/src/pages/Admin/Publications.vue [hunk 2]
```diff
@@ -95,7 +105,17 @@ export default {
...
+  watch: {
+    "new_publication.name": function() {
+      this.vuelidateExternalResults.new_publication.name = [];
+    }
+  },
   methods: {
+    getErrorMessageKey,
+    resetForm() {
+      this.new_publication.name = "";
+        this.$v.$reset();
```
Comment (gmeben) [inline client/src/pages/Admin/Publications.vue:117]: Is this function call necessary? If it is, is this indentation intentional?

**Review by gmeben — CHANGES_REQUESTED** [review]
I managed to get halfway through the files and found some things of note. I'll continue my review tomorrow and hopefully finish then.

File: client/src/components/molecules/ErrorFieldRenderer.vue [hunk 3]
```diff
@@ -0,0 +1,33 @@
+<template>
+  <div>
+    <div
+      v-for="error in errors"
+      :key="error.$validator"
+      data-cy="name_field_error"
+    >
+      {{ $t(`${prefix}.${getErrorMessageKey(error)}`) }}
+    </div>
+  </div>
```
[the remark this answers is not in the document]
Comment (wreality) [reply client/src/components/molecules/ErrorFieldRenderer.vue:10]: meh, article isn't really semantic... how about a ul/li ?

File: client/src/pages/Admin/Publications.vue (hunk 1 above)
Comment (wreality) [reply client/src/pages/Admin/Publications.vue:26]: Quasar has no convention.  It just gives you a slot that you can fill with whatever you like.  Vuelidate evaluates all of the rules independently so each field could have more than one error.  Seems better to show them to the user (if they all fit comfortably) than to try to pick which error to show them.

And yes, the error field renderer should be in a slot.  I'll refactor.

File: client/src/pages/Admin/Publications.vue (hunk 2 above)
Comment (wreality) [reply client/src/pages/Admin/Publications.vue:117]: The indentation is indeed wrong. I'll fix.

As to the necessity of the function call, I just generally like to create a reset form function so that as more things have to be done they have a place to go.  Also makes the handle method a bit more readable. IMHO.

File: client/src/components/molecules/ErrorFieldRenderer.vue (hunk 3 above)
[the remark this answers is not in the document]
Comment (gmeben) [reply client/src/components/molecules/ErrorFieldRenderer.vue:10]: I would argue `article` is semantic for error messages as they can be [independent items of content that are meant to be reused](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/article). `article` also semantically gives error messages greater importance in the document outline to better honor the attention they demand to be brought to them.

If you don't find my argument compelling then an unordered list is fine. Just be sure to style them accordingly.

File: client/src/pages/Admin/Publications.vue (hunk 2 above)
Comment (gmeben) [reply client/src/pages/Admin/Publications.vue]: No, I meant: is the call to `this.$v.$reset();` necessary inside the reset function?

File: client/src/components/molecules/ErrorFieldRenderer.vue (hunk 3 above)
[the remark this answers is not in the document]
Comment (wreality) [reply client/src/components/molecules/ErrorFieldRenderer.vue]: Ok, changed to ul.  Looked at the link you provided regarding article and I'm not sold this would be an appropriate use.

File: client/src/pages/Admin/Publications.vue (hunk 2 above)
Comment (wreality) [reply client/src/pages/Admin/Publications.vue]: Yep, you're right that's an artifact from a previous implementation.  Removed.

**Comment by wreality** [comment]
I see there's a failing test now. I'm working on getting it to pass again.

File: client/src/use/user.js [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+      update: (cache, { data: { login } }) => {
+        cache.writeQuery({
+          query: CURRENT_USER,
+          data: { currentUser: { ...login } },
+        })
+      },
+    }))
+
+    /**
+     * Login the suppled user
```
Comment (gmeben) [inline client/src/use/user.js:75]: Typo
```suggestion
     * Login the supplied user
```

File: client/src/use/userValidation.jest.spec.js [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+        expect(eV.email.$invalid).toBeFalsy();
+        expect(eV.required.$invalid).toBeFalsy();
+        expect(eV.$invalid).toBeFalsy();
+
+    });
+
+    text('locally validated password', () => {
+        const { result: { $v, user } } = mountComposable();
+        const eV = $v.value.password;
+
```
Comment (gmeben) [inline client/src/use/userValidation.jest.spec.js:57]: Extra newline
```suggestion
```

File: client/src/use/userValidation.js [hunk 6]
```diff
@@ -0,0 +1,71 @@
+import { reactive, watch } from '@vue/composition-api';
+import useVuelidate from "@vuelidate/core";
+import { required, email } from "@vuelidate/validators";
+import { CREATE_USER } from "src/graphql/mutations";
+import { useMutation } from "@vue/apollo-composable";
+import zxcvbn from "zxcvbn";
+import { applyExternalValidationErrors } from 'src/use/validationHelpers';
+
```
Comment (gmeben) [inline client/src/use/userValidation.js:8]: Extra newline
```suggestion
```

File: client/src/use/validationHelpers.js [hunk 7]
```diff
@@ -0,0 +1,115 @@
+import { computed, watch, inject } from '@vue/composition-api';
+import { clone } from 'lodash';
+
+/**
+ * Create a computed property for checking for and error condition on a field.
```
Comment (gmeben) [inline client/src/use/validationHelpers.js:5]: Typo
```suggestion
 * Create a computed property for checking for an error condition on a field.
```

File: client/src/use/validationHelpers.js [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+    return computed(() => {
+
+       return (field, key) => {
+            return hasErrorKey(validator.value?.[field].$errors, key) ?? false
+        }
+    })
+}
+
+
```
Comment (gmeben) [inline client/src/use/validationHelpers.js:23]: Extra newline
```suggestion
```

File: client/src/use/validationHelpers.js [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+/**
+ * Checks the supplied validation errors array for the presence of a validator
+ * or a externalResults message.
```
Comment (gmeben) [inline client/src/use/validationHelpers.js:26]: Typo
```suggestion
 * or an externalResults message.
```

File: client/src/use/validationHelpers.js [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+
+/**
+ * Add a one-shot watcher to a reactive property field that removes itself after
+ * being called once when the property values changes.
+ *
+ * @param {reactive} data
+ * @param {reactive} property
+ * @param {Function} callback
+ */
+
```
Comment (gmeben) [inline client/src/use/validationHelpers.js:73]: Extra newline
```suggestion
```

File: client/src/use/validationHelpers.js [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+            }
+        }
+    )
+}
+
+/**
+ * Parses a graphql error object for any validation errors and applies them to the
+ * provided externalValidation reactive.  Apply the externalValidation reactive to
+ * vuelidates $externalResults option to include graphql validation errors in vuelidate
+ * error responses.
```
Comment (gmeben) [inline client/src/use/validationHelpers.js:90]: Typos
```suggestion
 * Parses a GraphQL error object for any validation errors and applies them to the
 * provided externalValidation reactive.  Apply the externalValidation reactive to
 * vuelidate's $externalResults option to include GraphQL validation errors in 
 * vuelidate error responses.
```

File: client/src/use/validationHelpers.js [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+ *
+ * @param {reactive} data Form data reactive
+ * @param {reactive} externalValidation External validation reactive
+ * @param {Object} error Error object returned by ApolloClient
+ * @param {String} strip String to strip from the beginning of the graphql field name
```
Comment (gmeben) [inline client/src/use/validationHelpers.js:95]:
```suggestion
 * @param {String} strip String to strip from the beginning of the GraphQL field name
```

File: client/test/cypress/integration/admin/publications.spec.js [hunk 13]
```diff
@@ -25,36 +26,34 @@ describe('Admin Publications', () => {
...
-    cy.injectAxe();
+    cy.dataCy('name_field_error').should('be.visible');
     cy.dataCy('new_publication_input')
       .type('Draft Publication from Cypress');
     cy.checkA11y();
   });
 
   it('prevents publication creation when the name exceeds the maximum length', () => {
-    const name_256_characters = '0123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345';
+    const name_256_characters = '01234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123450{enter}';
```
Comment (gmeben) [inline client/test/cypress/integration/admin/publications.spec.js:36]: The `{enter}` is already present on line 38 and the addition of the `0` character makes this a 257 character string. Either rename the variable to `name_257_characters` and change line 38 or revert this change as suggested.
```suggestion
    const name_256_characters = '0123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345';
```

File: client/test/cypress/integration/admin/publications.spec.js [hunk 14]
```diff
@@ -25,36 +26,34 @@ describe('Admin Publications', () => {
...
+    cy.get('.q-transition--field-message-leave-active').should('not.exist');
     cy.checkA11y();
   });
 
   it('prevents publication creation when the name is not unique', () => {
     cy.dataCy('new_publication_input')
       .type('Duplicate Publication from Cypress{enter}');
-    cy.dataCy('create_publication_notify').should('be.visible').should('have.class','bg-positive');
-    cy.injectAxe();
+    cy.dataCy('create_publication_notify').should('be.visible');
```
Comment (gmeben) [inline client/test/cypress/integration/admin/publications.spec.js:47]: Wouldn't it be a stronger test to assert the type of the notification that's expected in addition to the notification being visible?
```suggestion
    cy.dataCy('create_publication_notify').should('be.visible').should('have.class','bg-positive');
```

File: client/test/cypress/integration/register.spec.js [hunk 15]
```diff
@@ -77,6 +63,28 @@ describe("Register", () => {
                 .parents('label')
                 .should('not.have.class', 'q-field--error')
 
+             //Username must be unique
```
Comment (gmeben) [inline client/test/cypress/integration/register.spec.js:66]: Extra space
```suggestion
            //Username must be unique
```

File: client/test/cypress/integration/register.spec.js [hunk 16]
```diff
@@ -77,6 +63,28 @@ describe("Register", () => {
...
+            cy.dataCy('username_field')
+                .clear()
+                .type('regularUser{enter}')
+                .parents('label')
+                .should('have.class', 'q-field--error')
+                 .contains('is not available');
+            cy.dataCy('username_field')
+                .type('brandnewusername')
+
+             //Email must be unique
```
Comment (gmeben) [inline client/test/cypress/integration/register.spec.js:76]: Extra space
```suggestion
            //Email must be unique
```

File: client/test/cypress/integration/register.spec.js [hunk 17]
```diff
@@ -77,6 +63,28 @@ describe("Register", () => {
...
+            cy.dataCy('email_field')
+             .clear()
+             .type('regularuser@ccrproject.dev{enter}')
+             .parents('label')
+             .should('have.class', 'q-field--error')
+                 .contains('already registered');
```
Comment (gmeben) [inline client/test/cypress/integration/register.spec.js:82]: Inconsistent indentation
```suggestion
            cy.dataCy('email_field')
                .clear()
                .type('regularuser@ccrproject.dev{enter}')
                .parents('label')
                .should('have.class', 'q-field--error')
                .contains('already registered');
```

File: client/test/cypress/integration/register.spec.js [hunk 18]
```diff
@@ -77,6 +63,28 @@ describe("Register", () => {
                 .parents('label')
                 .should('not.have.class', 'q-field--error')
 
+             //Username must be unique
+            cy.dataCy('username_field')
+                .clear()
+                .type('regularUser{enter}')
+                .parents('label')
+                .should('have.class', 'q-field--error')
+                 .contains('is not available');
```
Comment (gmeben) [inline client/test/cypress/integration/register.spec.js:72]: Extra space
```suggestion
                .contains('is not available');
```

**Review by gmeben — CHANGES_REQUESTED** [review]

**Comment by wreality** [comment]
Geez I'm embarrassed by all the typos. Sorry about that.

File: client/test/cypress/integration/admin/publications.spec.js (hunk 13 above)
Comment (wreality) [reply client/test/cypress/integration/admin/publications.spec.js:36]: I need to investigate the underlying reason why I had to do this to get the test to pass.

File: client/test/cypress/integration/admin/publications.spec.js (hunk 14 above)
Comment (wreality) [reply client/test/cypress/integration/admin/publications.spec.js:47]: You're right, but I had a reason why I changed it.  I'll investigate why I did that and update.

File: client/test/cypress/integration/admin/publications.spec.js (hunk 13 above)
Comment (wreality) [reply client/test/cypress/integration/admin/publications.spec.js:36]: Updated.

**Review by gmeben — APPROVED** [review]
